## Content Analysis and identifying categories with the largest popularity

In [731]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

**There are total 3 datasets relevant to the analysis.**

#### 1. Content

In [732]:
df_content = pd.read_csv("Content.csv")
df_content.head()

Unnamed: 0                            Content ID  \
0           0  97522e57-d9ab-4bd6-97bf-c24d952602d2   
1           1  9f737e0a-3cdd-4d29-9d24-753f4e3be810   
2           2  230c4e4d-70c3-461d-b42c-ec09396efb3f   
3           3  356fff80-da4d-4785-9f43-bc1261031dc6   
4           4  01ab84dd-6364-4236-abbb-3f237db77180   

                                User ID   Type        Category  \
0  8d3cd87d-8a31-4935-9a4f-b319bfe05f31  photo        Studying   
1  beb1f34e-7870-46d6-9fc7-2e12eb83ce43  photo  healthy eating   
2  a5c65404-5894-4b87-82f2-d787cbee86b4  photo  healthy eating   
3  9fb4ce88-fac1-406c-8544-1a899cee7aaf  photo      technology   
4  e206e31b-5f85-4964-b6ea-d7ee5324def1  video            food   

                                                 URL  
0  https://socialbuzz.cdn.com/content/storage/975...  
1  https://socialbuzz.cdn.com/content/storage/9f7...  
2  https://socialbuzz.cdn.com/content/storage/230...  
3  https://socialbuzz.cdn.com/content/storage/356...  
4  https://socialbuzz.cdn.com/content/storage/01a...

In [733]:
df_content.drop(columns = ['Unnamed: 0'], inplace = True)

In [734]:
df_content.head(2)

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2  8d3cd87d-8a31-4935-9a4f-b319bfe05f31   
1  9f737e0a-3cdd-4d29-9d24-753f4e3be810  beb1f34e-7870-46d6-9fc7-2e12eb83ce43   

    Type        Category                                                URL  
0  photo        Studying  https://socialbuzz.cdn.com/content/storage/975...  
1  photo  healthy eating  https://socialbuzz.cdn.com/content/storage/9f7...

In [735]:
df_content.shape

(1000, 5)

In [736]:
df_content.columns

Index(['Content ID', 'User ID', 'Type', 'Category', 'URL'], dtype='object')

In [737]:
df_content.isna().sum()

Content ID      0
User ID         0
Type            0
Category        0
URL           199
dtype: int64

There are 199 null values in the URL column, as per the requirement, the client wants to drop all the rows containing null values, thus, dropping them.

In [738]:
df_content.dropna(inplace=True)

In [739]:
df_content.isna().sum()

Content ID    0
User ID       0
Type          0
Category      0
URL           0
dtype: int64

So, now no null values present in the data.

In [740]:
df_content.Category.value_counts()

Category
travel             60
science            55
fitness            54
animals            54
culture            54
technology         54
tennis             53
cooking            51
healthy eating     50
dogs               49
education          48
soccer             47
food               47
veganism           43
studying           41
public speaking    40
Studying            1
Name: count, dtype: int64

In [741]:
df_content['Category'] = df_content['Category'].replace('Studying', 'studying')
df_content.Category.value_counts()

Category
travel             60
science            55
technology         54
animals            54
fitness            54
culture            54
tennis             53
cooking            51
healthy eating     50
dogs               49
education          48
food               47
soccer             47
veganism           43
studying           42
public speaking    40
Name: count, dtype: int64

In [742]:
df_content.rename(columns = {'Type': 'Content Type'}, inplace=True)
df_content.head(2)

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2  8d3cd87d-8a31-4935-9a4f-b319bfe05f31   
1  9f737e0a-3cdd-4d29-9d24-753f4e3be810  beb1f34e-7870-46d6-9fc7-2e12eb83ce43   

  Content Type        Category  \
0        photo        studying   
1        photo  healthy eating   

                                                 URL  
0  https://socialbuzz.cdn.com/content/storage/975...  
1  https://socialbuzz.cdn.com/content/storage/9f7...

In [743]:
df_content.dtypes

Content ID      object
User ID         object
Content Type    object
Category        object
URL             object
dtype: object

#### 2. Reactions

In [744]:
df_reactions = pd.read_csv('Reactions.csv')
df_reactions.head(2)

Unnamed: 0                            Content ID  \
0           0  97522e57-d9ab-4bd6-97bf-c24d952602d2   
1           1  97522e57-d9ab-4bd6-97bf-c24d952602d2   

                                User ID     Type             Datetime  
0                                   NaN      NaN  2021-04-22 15:17:15  
1  5d454588-283d-459d-915d-c48a2cb4c27f  disgust  2020-11-07 09:43:50

In [745]:
df_reactions.drop(columns = ['Unnamed: 0'], inplace = True)
df_reactions.head(2)

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2                                   NaN   
1  97522e57-d9ab-4bd6-97bf-c24d952602d2  5d454588-283d-459d-915d-c48a2cb4c27f   

      Type             Datetime  
0      NaN  2021-04-22 15:17:15  
1  disgust  2020-11-07 09:43:50

In [746]:
df_reactions.shape

(25553, 4)

In [747]:
df_reactions.columns

Index(['Content ID', 'User ID', 'Type', 'Datetime'], dtype='object')

In [748]:
df_reactions.isna().sum()

Content ID       0
User ID       3019
Type           980
Datetime         0
dtype: int64

There are 3019 null values in UserId and 980 in the Type columns, thus, dropping those columns.

In [749]:
df_reactions.dropna(inplace=True)

In [750]:
df_reactions.isna().sum()

Content ID    0
User ID       0
Type          0
Datetime      0
dtype: int64

All null values cleaned

In [751]:
df_reactions['Content ID'].nunique()

962

In [752]:
df_content['Content ID'].nunique()

801

Now, there is a discrepancy between the no. of unique content Ids in content table and the reactions table. Contents table has 801 unique contents, while reactions has 962.

In [753]:
df_reactions.dtypes

Content ID    object
User ID       object
Type          object
Datetime      object
dtype: object

Here, Datetime is object, however it should be in datetime format. Thus, changing the format for it.

In [754]:
df_reactions['Datetime'] = pd.to_datetime(df_reactions['Datetime'], errors='coerce')
df_reactions.dtypes

Content ID            object
User ID               object
Type                  object
Datetime      datetime64[ns]
dtype: object

The data type of Datetime column is now converted successfully.

In [755]:
df_reactions.rename(columns = {'Type': 'Reaction Type'}, inplace=True)
df_reactions.head(2)

Content ID                               User ID  \
1  97522e57-d9ab-4bd6-97bf-c24d952602d2  5d454588-283d-459d-915d-c48a2cb4c27f   
2  97522e57-d9ab-4bd6-97bf-c24d952602d2  92b87fa5-f271-43e0-af66-84fac21052e6   

  Reaction Type            Datetime  
1       disgust 2020-11-07 09:43:50  
2       dislike 2021-06-17 12:22:51

#### 3. df_reactionTypes

In [756]:
df_reactionTypes = pd.read_csv('ReactionTypes.csv')
df_reactionTypes.head(2)

Unnamed: 0   Type Sentiment  Score
0           0  heart  positive     60
1           1   want  positive     70

In [757]:
df_reactionTypes.drop(columns = ['Unnamed: 0'], inplace = True)
df_reactionTypes.head(2)

Type Sentiment  Score
0  heart  positive     60
1   want  positive     70

In [758]:
df_reactionTypes.columns

Index(['Type', 'Sentiment', 'Score'], dtype='object')

In [759]:
df_reactionTypes.shape

(16, 3)

In [760]:
df_reactionTypes.isna().sum()

Type         0
Sentiment    0
Score        0
dtype: int64

In [761]:
df_reactionTypes.dtypes

Type         object
Sentiment    object
Score         int64
dtype: object

In [762]:
df_reactionTypes.rename(columns = {'Type': 'Reaction Type'}, inplace=True)
df_reactionTypes.head(2)

Reaction Type Sentiment  Score
0         heart  positive     60
1          want  positive     70

### Identifying the relevant columns

In [763]:
df_content.columns

Index(['Content ID', 'User ID', 'Content Type', 'Category', 'URL'], dtype='object')

Here, the URL column doesn't seem very useful for the analysis of popular categories, thus, dropping it.

In [764]:
df_content.drop(columns = ['URL'], inplace = True)
df_content.columns

Index(['Content ID', 'User ID', 'Content Type', 'Category'], dtype='object')

In [765]:
df_reactions.columns

Index(['Content ID', 'User ID', 'Reaction Type', 'Datetime'], dtype='object')

Here, we will keep all the columns considering we can analyze at what time which content is favoured by what number of users.

In [766]:
df_reactionTypes.columns

Index(['Reaction Type', 'Sentiment', 'Score'], dtype='object')

Here as well, all the columns will be retained for now.

## Merging the final file:

In [767]:
df_content.head(2)

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2  8d3cd87d-8a31-4935-9a4f-b319bfe05f31   
1  9f737e0a-3cdd-4d29-9d24-753f4e3be810  beb1f34e-7870-46d6-9fc7-2e12eb83ce43   

  Content Type        Category  
0        photo        studying  
1        photo  healthy eating

In [768]:
df_content.shape

(801, 4)

In [769]:
df_content['Content ID'].nunique()

801

There are 801 unique content Ids here.

In [770]:
df_content['User ID'].nunique()

408

So, this shows, more than 1 user is consuming by multiple contents.

So, we will create a df of user Id and no. of contents watched.

In [771]:
df_viewer = df_content.groupby('User ID')['Content ID'].nunique().reset_index()
df_viewer.columns = ['User  ID', 'No. of Content Viewed']
df_viewer.head()

User  ID  No. of Content Viewed
0  002cd824-10f2-447c-8d1c-940325a1cdf4                      2
1  007a1366-a46f-40f1-9ba5-01245aeaf20f                      2
2  04e52602-e370-4a99-bfda-f0c1655058da                      1
3  054b5d2a-7ee3-49c8-912a-1cf58389ae60                      2
4  05782669-264e-4c14-990e-14c8dbb13bef                      5

In [772]:
df_viewer.shape

(408, 2)

In [773]:
df_viewer['No. of Content Viewed'].value_counts()

No. of Content Viewed
1    161
2    149
3     62
4     28
5      5
6      2
7      1
Name: count, dtype: int64

Well, it is clear from the above data that one user is viewing multiple content, however, one content is currently recorded to be viewed only by one person. Thus, this User ID column will not be considered in further analysis of popular content.

In [774]:
df_content.drop(columns = ['User ID'], inplace = True)

In [775]:
df_reactions.head(2)

Content ID                               User ID  \
1  97522e57-d9ab-4bd6-97bf-c24d952602d2  5d454588-283d-459d-915d-c48a2cb4c27f   
2  97522e57-d9ab-4bd6-97bf-c24d952602d2  92b87fa5-f271-43e0-af66-84fac21052e6   

  Reaction Type            Datetime  
1       disgust 2020-11-07 09:43:50  
2       dislike 2021-06-17 12:22:51

In [776]:
df_reactions['Datetime'].value_counts()

Datetime
2020-10-29 20:51:08    2
2020-09-10 06:59:59    2
2020-11-01 00:05:07    2
2020-09-11 05:52:04    2
2020-12-13 17:37:25    2
                      ..
2020-08-17 07:47:08    1
2021-02-15 19:24:31    1
2020-07-03 02:35:26    1
2020-11-15 04:11:51    1
2021-01-04 04:55:11    1
Name: count, Length: 22524, dtype: int64

In [777]:
content_map = df_content.set_index('Content ID')['Category']

# Map Category to df_reactions
df_reactions['Category'] = df_reactions['Content ID'].map(content_map)
df_reactions.head()

Content ID                               User ID  \
1  97522e57-d9ab-4bd6-97bf-c24d952602d2  5d454588-283d-459d-915d-c48a2cb4c27f   
2  97522e57-d9ab-4bd6-97bf-c24d952602d2  92b87fa5-f271-43e0-af66-84fac21052e6   
3  97522e57-d9ab-4bd6-97bf-c24d952602d2  163daa38-8b77-48c9-9af6-37a6c1447ac2   
4  97522e57-d9ab-4bd6-97bf-c24d952602d2  34e8add9-0206-47fd-a501-037b994650a2   
5  97522e57-d9ab-4bd6-97bf-c24d952602d2  9b6d35f9-5e15-4cd0-a8d7-b1f3340e02c4   

  Reaction Type            Datetime  Category  
1       disgust 2020-11-07 09:43:50  studying  
2       dislike 2021-06-17 12:22:51  studying  
3        scared 2021-04-18 05:13:58  studying  
4       disgust 2021-01-06 19:13:01  studying  
5    interested 2020-08-23 12:25:58  studying

In [778]:
df_reactions.Category.value_counts()

Category
travel             1368
science            1351
animals            1323
healthy eating     1303
cooking            1277
culture            1217
food               1183
tennis             1153
technology         1150
education          1127
soccer             1065
fitness            1050
dogs               1047
veganism            948
studying            911
public speaking     911
Name: count, dtype: int64

In [779]:
df_reactions.isna().sum()

Content ID          0
User ID             0
Reaction Type       0
Datetime            0
Category         4150
dtype: int64

In [780]:
df_reactions['Category'] = df_reactions['Category'].fillna('miscellaneous')
df_reactions.isna().sum()

Content ID       0
User ID          0
Reaction Type    0
Datetime         0
Category         0
dtype: int64

In [781]:
df_reactions['Content ID'][df_reactions['Category'] == 'miscellaneous'].nunique()

190

So there are 190 Content Ids which have now been labelled as Miscelleanous category, else if dropped these would have been lost.

In [782]:
sentiment_map = df_reactionTypes.set_index('Reaction Type')['Score']

# Map Category to df_reactions
df_reactions['Sentiment_score'] = df_reactions['Reaction Type'].map(sentiment_map)
df_reactions.head()

Content ID                               User ID  \
1  97522e57-d9ab-4bd6-97bf-c24d952602d2  5d454588-283d-459d-915d-c48a2cb4c27f   
2  97522e57-d9ab-4bd6-97bf-c24d952602d2  92b87fa5-f271-43e0-af66-84fac21052e6   
3  97522e57-d9ab-4bd6-97bf-c24d952602d2  163daa38-8b77-48c9-9af6-37a6c1447ac2   
4  97522e57-d9ab-4bd6-97bf-c24d952602d2  34e8add9-0206-47fd-a501-037b994650a2   
5  97522e57-d9ab-4bd6-97bf-c24d952602d2  9b6d35f9-5e15-4cd0-a8d7-b1f3340e02c4   

  Reaction Type            Datetime  Category  Sentiment_score  
1       disgust 2020-11-07 09:43:50  studying                0  
2       dislike 2021-06-17 12:22:51  studying               10  
3        scared 2021-04-18 05:13:58  studying               15  
4       disgust 2021-01-06 19:13:01  studying                0  
5    interested 2020-08-23 12:25:58  studying               30

In [783]:
df_reactions.groupby('Content ID')['Sentiment_score'].median().reset_index()

Content ID  Sentiment_score
0    004e820e-49c3-4ba2-9d02-62db0065410c             60.0
1    00d0cdf9-5919-4102-bf84-ebde253c3cd2             45.0
2    01396602-c759-4a17-90f0-8f9b3ca11b30             25.0
3    019b61f4-926c-438e-adaf-6119c5eab752             30.0
4    01ab84dd-6364-4236-abbb-3f237db77180             15.0
..                                    ...              ...
957  fdca8d15-966b-4825-8133-1fafc5c1f9fc             50.0
958  fe06b730-b1f8-4f55-af1a-52487d8f1ec6             45.0
959  fea8d77c-fd0b-4678-868f-fbae567642f3             60.0
960  fea9077f-2fe7-43bd-aaef-dc2619988d94             32.5
961  ff883828-a610-492d-8635-8a777eaad25f             35.0

[962 rows x 2 columns]

In [784]:
from functools import reduce
# Calculate unique user counts
unique_user_counts = df_reactions.groupby('Content ID')['User ID'].nunique().reset_index()
unique_user_counts.columns = ['Content ID', 'Unique User Count']

# Calculate unique reaction counts
unique_reaction_counts = df_reactions.groupby('Content ID')['Reaction Type'].nunique().reset_index()
unique_reaction_counts.columns = ['Content ID', 'Unique Reaction Count']

unique_reaction_sentiment = df_reactions.groupby('Content ID')['Sentiment_score'].mean().reset_index()
unique_reaction_sentiment.columns = ['Content ID', 'Mean Reaction Sentiment']

def custom_median(series):
    sorted_values = sorted(series)
    n = len(sorted_values)
    if n % 2 == 1:
        return sorted_values[n // 2]  # Odd case: return the middle value
    else:
        return min(sorted_values[n // 2 - 1], sorted_values[n // 2])  # Even case: return the larger of the two middle values

# Calculate median reaction sentiment
median_reaction_sentiment = df_reactions.groupby('Content ID')['Sentiment_score'].agg(custom_median).reset_index()
median_reaction_sentiment.columns = ['Content ID', 'Median Reaction Sentiment']

# Merge both DataFrames on 'Content ID'
df_unique_content_viewers = reduce(lambda left, right: pd.merge(left, right, on='Content ID', how='left'),
                                    [unique_user_counts, unique_reaction_counts, unique_reaction_sentiment,median_reaction_sentiment])
df_unique_content_viewers.head()

Content ID  Unique User Count  \
0  004e820e-49c3-4ba2-9d02-62db0065410c                  1   
1  00d0cdf9-5919-4102-bf84-ebde253c3cd2                 40   
2  01396602-c759-4a17-90f0-8f9b3ca11b30                 35   
3  019b61f4-926c-438e-adaf-6119c5eab752                 12   
4  01ab84dd-6364-4236-abbb-3f237db77180                  1   

   Unique Reaction Count  Mean Reaction Sentiment  Median Reaction Sentiment  
0                      1                60.000000                         60  
1                     16                41.333333                         45  
2                     14                29.333333                         20  
3                      8                33.666667                         30  
4                      1                15.000000                         15

In [785]:
reaction_map = df_reactionTypes.set_index('Score')['Sentiment']
reaction_map.to_dict()

{60: 'positive',
 70: 'positive',
 0: 'negative',
 5: 'negative',
 30: 'positive',
 20: 'neutral',
 65: 'positive',
 75: 'positive',
 72: 'positive',
 50: 'positive',
 10: 'negative',
 45: 'positive',
 35: 'neutral',
 15: 'negative',
 12: 'negative'}

In [786]:
# Create a function to map sentiment scores to sentiment names
def map_sentiment(score):
    for sentiment_score, sentiment_name in reaction_map.items():
        if score == sentiment_score:
            return sentiment_name

# Apply the function to the 'Median Reaction Sentiment' column
df_unique_content_viewers['Overall Sentiment'] = df_unique_content_viewers['Median Reaction Sentiment'].apply(map_sentiment)
df_unique_content_viewers.head()

Content ID  Unique User Count  \
0  004e820e-49c3-4ba2-9d02-62db0065410c                  1   
1  00d0cdf9-5919-4102-bf84-ebde253c3cd2                 40   
2  01396602-c759-4a17-90f0-8f9b3ca11b30                 35   
3  019b61f4-926c-438e-adaf-6119c5eab752                 12   
4  01ab84dd-6364-4236-abbb-3f237db77180                  1   

   Unique Reaction Count  Mean Reaction Sentiment  Median Reaction Sentiment  \
0                      1                60.000000                         60   
1                     16                41.333333                         45   
2                     14                29.333333                         20   
3                      8                33.666667                         30   
4                      1                15.000000                         15   

  Overall Sentiment  
0          positive  
1          positive  
2           neutral  
3          positive  
4          negative

In [787]:
df_unique_content_viewers['Overall Sentiment'].value_counts()

Overall Sentiment
positive    585
neutral     318
negative     59
Name: count, dtype: int64

In [788]:
df_unique_content_viewers['Category'] = df_unique_content_viewers['Content ID'].map(content_map)
df_unique_content_viewers.head()

Content ID  Unique User Count  \
0  004e820e-49c3-4ba2-9d02-62db0065410c                  1   
1  00d0cdf9-5919-4102-bf84-ebde253c3cd2                 40   
2  01396602-c759-4a17-90f0-8f9b3ca11b30                 35   
3  019b61f4-926c-438e-adaf-6119c5eab752                 12   
4  01ab84dd-6364-4236-abbb-3f237db77180                  1   

   Unique Reaction Count  Mean Reaction Sentiment  Median Reaction Sentiment  \
0                      1                60.000000                         60   
1                     16                41.333333                         45   
2                     14                29.333333                         20   
3                      8                33.666667                         30   
4                      1                15.000000                         15   

  Overall Sentiment        Category  
0          positive          tennis  
1          positive  healthy eating  
2           neutral          tennis  
3          positive             NaN  
4          negative            food

In [789]:
df_unique_content_viewers.isna().sum()

Content ID                     0
Unique User Count              0
Unique Reaction Count          0
Mean Reaction Sentiment        0
Median Reaction Sentiment      0
Overall Sentiment              0
Category                     190
dtype: int64

In [790]:
df_unique_content_viewers['Category'] = df_unique_content_viewers['Category'].fillna('miscellaneous')
df_unique_content_viewers.isna().sum()

Content ID                   0
Unique User Count            0
Unique Reaction Count        0
Mean Reaction Sentiment      0
Median Reaction Sentiment    0
Overall Sentiment            0
Category                     0
dtype: int64

In [791]:
df_unique_content_viewers['Sentiment Difference'] = abs(df_unique_content_viewers['Mean Reaction Sentiment'] - df_unique_content_viewers['Median Reaction Sentiment'])

# Filter Content IDs where the difference is greater than 5
high_difference_content_ids = df_unique_content_viewers[df_unique_content_viewers['Sentiment Difference'] > 10]['Content ID']

# Display the result
print(len(high_difference_content_ids))

224


In [792]:
# Filter Content IDs where the difference is greater than 10
df_high_difference = df_unique_content_viewers[df_unique_content_viewers['Sentiment Difference'] > 10].copy()

# Display the result
df_high_difference.head()

Content ID  Unique User Count  \
18  0411a1da-97e1-44b6-945d-06b4682ba802                 28   
29  072fb6c2-1505-408a-a35d-f0c4ffaf2696                  4   
33  09bb0893-a05b-49bc-95ec-1e158577bf13                 24   
38  0adb9a94-210a-4e84-bca1-13db9452ff0e                 23   
39  0be329af-19d6-4a34-8bc1-f4391e9877dd                  4   

    Unique Reaction Count  Mean Reaction Sentiment  Median Reaction Sentiment  \
18                     13                47.321429                         60   
29                      3                51.250000                         65   
33                     12                49.346154                         60   
38                     10                46.478261                         60   
39                      4                33.750000                         20   

   Overall Sentiment        Category  Sentiment Difference  
18          positive      technology             12.678571  
29          positive         fitness             13.750000  
33          positive  healthy eating             10.653846  
38          positive   miscellaneous             13.521739  
39           neutral            dogs             13.750000

## Required merged df

In [793]:
df_unique_content_viewers.to_csv('df_unique_content_viewers.csv', index=False)

In [794]:
df_reactions.to_csv('df_reactions_with_datetime.csv',index = False)